### PR TITLE
Fix buffer size calculation in algorithm_impl_hetero.h

### PR DESCRIPTION
**Pull request by SergeyKopienko**
In this PR we fix case for remove_copy algorithm and other when output buffer size is less then source buffer size.
In this case we had assertion at [include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h](https://github.com/oneapi-src/oneDPL/blob/main/include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h), line 493:
```
assert(__offset + __n <= __size);
```

Example for reproduce this assert:
```
    dpct::device_ext& dev_ct1 = dpct::get_current_device();
    sycl::queue& q_ct1 = dev_ct1.default_queue(); // device iterator
    const int N = 6;
    int A[N] = {-2, 0, -1, 0, 1, 2};
    int B[N - 2];
    int ans[N - 2] = {-2, -1, 1, 2};
    dpct::device_vector<int> V(A, A + N);
    dpct::device_vector<int> result(B, B + N - 2);

    oneapi::dpl::remove_copy(oneapi::dpl::execution::make_device_policy(q_ct1), V.begin(), V.end(), result.begin(), 0);

```
In this example the size of output buffer ```result``` is less then the size of source buffer ```V``` and we have failed statement in assert.
But there are no requirements for this algorithm about output buffer size at [cppreference](https://en.cppreference.com/w/cpp/algorithm/remove_copy).

This issue was introduced in PR [Fix assert in reduce_by_segment_impl #573](https://github.com/oneapi-src/oneDPL/pull/573)

### CODE REVIEW

**Comment by SergeyKopienko** [comment]
@MikeDvorskiy this assert was introduced here: https://github.com/oneapi-src/oneDPL/pull/573#discussion_r906039552

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h [hunk 1]
```diff
@@ -94,7 +113,7 @@ __pattern_walk2(_ExecutionPolicy&& __exec, _ForwardIterator1 __first1, _ForwardI
     auto __buf1 = __keep1(__first1, __last1);
 
     auto __keep2 = oneapi::dpl::__ranges::__get_sycl_range<__acc_mode2, _ForwardIterator2>();
-    auto __buf2 = __keep2(__first2, __first2 + __n);
+    auto __buf2 = __keep2(__first2, __first2 + get_buffer_size()(__first2, __n));
```
Comment (MikeDvorskiy) [inline include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:116]: I think we have to make buffer size request within `__ranges::__get_sycl_range` class.
That class takes into account a transformation pipeline over a passed  hetero iterator.

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:116]: ok, will do

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h]: Fixed

**Pull request closed without merging**